flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java

```diff
@@ -82,8 +87,7 @@
 	/** The serializer for the type. May be eagerly initialized in the constructor,
 	 * or lazily once the {@link #initializeSerializerUnlessSet(ExecutionConfig)} method
 	 * is called. */
-	@Nullable
-	protected TypeSerializer<T> serializer;
+	private final AtomicReference<TypeSerializer<T>> serializerAtomicReference = new AtomicReference<>();
```

Comment:
Should `queryableStateName` and `ttlConfig` also be protected as they are also mutable fields?

Reply:
Yes if we aim at making `StateDescriptor` thread safe but that would be a bigger project and requires more testing.

Considering the seemingly fact that reset `queryableStateName` or `ttlConfig` won't cause obvious trouble, does it make sense if we fix the serializer lazy initialization issue (which has caused a NPE problem on kafka producer) first and then check whether it worth another JIRA/PR for making the whole `StateDescriptor` thread safe? Thanks.

Reply:
Sure, sounds good to me.